File: backend/backend/app/schemas/nutrition.py
import uuid
from datetime import datetime
from typing import Optional

from pydantic import BaseModel


# ===== CATEGORY SCHEMAS =====
class CategoryBase(BaseModel):
    name_fr: str
    name_en: str
    name_ar: str


class CategoryCreate(CategoryBase):
    pass


class CategoryUpdate(CategoryBase):
    name_fr: Optional[str] = None
    name_en: Optional[str] = None
    name_ar: Optional[str] = None


class CategoryPublic(CategoryBase):
    id: uuid.UUID
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class CategoriesPublic(BaseModel):
    data: list[CategoryPublic]
    count: int


# ===== INGREDIENT SCHEMAS =====
class IngredientBase(BaseModel):
    category_id: uuid.UUID
    name_fr: str
    name_en: str
    name_ar: str
    calories_per_100g: float
    protein_per_100g: float
    carbohydrates_per_100g: float
    fat_per_100g: float
    fiber_per_100g: float = 0.0
    sodium_per_100g: float = 0.0
    measurement_units: list[str] = []
    unit_conversions: dict[str, float] = {}
    tags: list[str] = []
    default_unit: str = "g"
    typical_serving: float = 100.0
    is_traditional: bool = False
    is_halal: bool = True


class IngredientCreate(IngredientBase):
    pass


class IngredientUpdate(BaseModel):
    category_id: Optional[uuid.UUID] = None
    name_fr: Optional[str] = None
    name_en: Optional[str] = None
    name_ar: Optional[str] = None
    calories_per_100g: Optional[float] = None
    protein_per_100g: Optional[float] = None
    carbohydrates_per_100g: Optional[float] = None
    fat_per_100g: Optional[float] = None
    fiber_per_100g: Optional[float] = None
    sodium_per_100g: Optional[float] = None
    measurement_units: Optional[list[str]] = None
    unit_conversions: Optional[dict[str, float]] = None
    tags: Optional[list[str]] = None
    default_unit: Optional[str] = None
    typical_serving: Optional[float] = None
    is_traditional: Optional[bool] = None
    is_halal: Optional[bool] = None


class IngredientPublic(IngredientBase):
    id: uuid.UUID
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class IngredientsPublic(BaseModel):
    data: list[IngredientPublic]
    count: int


# ===== MEAL INGREDIENT SCHEMAS =====
class MealIngredientBase(BaseModel):
    ingredient_id: uuid.UUID
    quantity: float
    unit: str
    calories: float = 0.0
    protein: float = 0.0
    carbohydrates: float = 0.0
    fat: float = 0.0
    fiber: float = 0.0
    sodium: float = 0.0


class MealIngredientCreate(MealIngredientBase):
    pass


class MealIngredientPublic(MealIngredientBase):
    id: uuid.UUID
    meal_id: uuid.UUID
    ingredient: Optional[IngredientPublic] = None

    class Config:
        from_attributes = True


# ===== MEAL SCHEMAS =====
class MealBase(BaseModel):
    name: str
    description: Optional[str] = None
    servings: float = 1.0
    total_calories: float = 0.0
    total_protein: float = 0.0
    total_carbohydrates: float = 0.0
    total_fat: float = 0.0
    total_fiber: float = 0.0
    total_sodium: float = 0.0
    is_favorite: bool = False
    is_traditional: bool = False


class MealCreate(MealBase):
    ingredients: list[MealIngredientCreate] = []


class MealUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    servings: Optional[float] = None
    total_calories: Optional[float] = None
    total_protein: Optional[float] = None
    total_carbohydrates: Optional[float] = None
    total_fat: Optional[float] = None
    total_fiber: Optional[float] = None
    total_sodium: Optional[float] = None
    is_favorite: Optional[bool] = None
    is_traditional: Optional[bool] = None
    ingredients: Optional[list[MealIngredientCreate]] = None


class MealPublic(MealBase):
    id: uuid.UUID
    user_id: Optional[uuid.UUID] = None
    created_at: datetime
    updated_at: datetime
    meal_ingredients: list[MealIngredientPublic] = []

    class Config:
        from_attributes = True


class MealsPublic(BaseModel):
    data: list[MealPublic]
    count: int


# ===== SAVED DAY SCHEMAS =====
class SavedDayBase(BaseModel):
    title: str
    description: Optional[str] = None
    day_plan: dict = {}


class SavedDayCreate(SavedDayBase):
    pass


class SavedDayUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    day_plan: Optional[dict] = None


class SavedDayPublic(SavedDayBase):
    id: uuid.UUID
    user_id: Optional[uuid.UUID] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class SavedDaysPublic(BaseModel):
    data: list[SavedDayPublic]
    count: int

File: frontend/lib/types/nutrition.ts
export interface Category {
  id: string
  name_fr: string
  name_en: string
  name_ar: string
}

export interface Ingredient {
  id: string
  category_id: string
  name_fr: string
  name_en: string
  name_ar: string
  calories_per_100g: number
  protein_per_100g: number
  carbohydrates_per_100g: number
  fat_per_100g: number
  fiber_per_100g: number
  sodium_per_100g: number
  measurement_units: string[]
  default_unit: string
  typical_serving: number
  unit_conversions: Record<string, number>
  tags: string[]
  is_traditional: boolean
  is_halal: boolean
}

export interface MealCategory {
  id: string
  name_fr: string
  name_en: string
  name_ar: string
  icon: string
  typical_time: string
  color: string
}

export interface MealIngredient {
  id: string
  ingredient: Ingredient
  quantity: number
  unit: string
  calories: number
  protein: number
  carbohydrates: number
  fat: number
  fiber: number
  sodium: number
}

export interface Meal {
  id?: string
  name: string
  description?: string
  ingredients: MealIngredient[]
  total_calories: number
  total_protein: number
  total_carbohydrates: number
  total_fat: number
  total_fiber: number
  total_sodium: number
  servings: number
  is_favorite: boolean
  is_traditional: boolean
}

export interface MealPlanItem {
  id: string
  type: "meal" | "ingredient"
  meal?: Meal
  ingredient?: Ingredient
  consumption_amount: number // grams for ingredients, servings or grams for meals
  consumption_unit?: "servings" | "grams" // Added unit tracking for meals
  consumption_percentage: number
  actual_calories: number
  actual_protein: number
  actual_carbohydrates: number
  actual_fat: number
  actual_fiber: number
  actual_sodium: number
}

export interface DailyPlan {
  id?: string
  user_id?: string
  plan_date: string
  meals: {
    breakfast?: MealPlanItem[]
    lunch?: MealPlanItem[]
    dinner?: MealPlanItem[]
    snacks?: MealPlanItem[]
  }
  total_calories: number
  total_protein: number
  total_carbohydrates: number
  total_fat: number
  total_fiber: number
  total_sodium: number
}

export interface WeeklyPlan {
  id?: string
  user_id?: string
  week_start: string
  daily_plans: DailyPlan[]
}

export interface NutritionGoals {
  daily_calories: number
  protein_percentage: number
  carbs_percentage: number
  fat_percentage: number
  fiber_grams: number
  sodium_mg: number
}

export interface DashboardStats {
  todayCalories: number
  calorieGoal: number
  weeklyAverage: number
  mealsPlanned: number
  favoriteMeals: number
  traditionalMealsCount: number
}

export interface UserProfile {
  id?: string
  user_id?: string
  name?: string
  preferred_language: "ar" | "fr" | "en"
  daily_calories: number
  protein_percentage: number
  carbs_percentage: number
  fat_percentage: number
  fiber_grams: number
  sodium_mg: number
  dietary_restrictions: string[]
  preferred_region?: string
}

export interface MealTypeSuggestion {
  name_en: string
  name_fr: string
  name_ar: string
  color: string
  typical_time: string
  icon: string
}

export interface DayMealItem {
  id: string
  type: "meal" | "ingredient"
  meal?: Meal
  meal_id?: string
  ingredient?: Ingredient
  ingredient_id?: string
  quantity: number
  unit: string
}

export interface DayMeal {
  id: string
  name: string
  time?: string
  color: string
  items: DayMealItem[]
  notes?: string
}

export interface DayPlan {
  date: string
  day_meals: DayMeal[]
  total_nutrition: {
    calories: number
    protein: number
    carbs: number
    fat: number
    fiber: number
    sodium: number
  }
  daily_goals: {
    calories: number
    protein: number
    carbs: number
    fat: number
    fiber: number
    sodium: number
  }
}

export interface Day {
  id: string
  title: string
  description?: string
  day_plan: DayPlan
  created_at: string
  updated_at: string
}
